Entity Events v1

i509VCB opened the pull request:
First up is an event fired after a living entity is damaged. All vanilla living entities except armor stands work with this event.

A few events include:
 - Entity changing world
 - Player changing world (seperate since these semantics are different)
 - Player respawn
 - Player copied (data half of respawn, is long before packets are sent)
 - When an entity is killed by another entity

This also fixes a bug related to vanilla's built-in killed by method not be invoked when a server player entity kills another entity.

i509VCB committed:
Entity Events v1.

First up is an event fired after a living entity is damaged. All vanilla living entities except armor stands work with this event.

liach commented:
Please explain why we need this damage callback.

i509VCB commented:
Could be used by mods that do logging of actions, may be useful for something like prism to track when an entity is attacked.

Also could be used with mods that effect the environment after an entity is damsged (I dunno visually make an entity show damage marks after being damaged).

dexman545 commented:
Iguana in a Blanket tracks entity health to modify attributes like speed.

liach commented:
Hmm, just curious, can you inject this where the stats are updated? I wonder if the stats are updated for armor stands at all. if armor stands have stats updated then it can be reliably detected by scoreboards hooked up with stats

i509VCB committed:
Events related to killing of entities, changing world, player respawn/copyFrom

Death related:
an entity killing something and an entity being killed by an adversary

Player related:
After respawn, copy to

i509VCB commented:
I don't think the armor stand touches stats, the only damage dealt stat incremented is for the player which isn't reliable for all damage types.

i509VCB committed:
Adversary stuff isn't needed

i509VCB committed:
checkstyle again

i509VCB committed:
Call AFTER_KILLED_OTHER on a ServerPlayerEntity upon death.

i509VCB committed:
Add event which is fired when a player first joined a server and add functional interface annotations

i509VCB committed:
Update build.gradle

i509VCB committed:
mispelled

liach commented on fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/EntityTeleportationHelper.java:
```diff
@@ -0,0 +1,117 @@
+/**
+ * A utility class providing methods to help teleport entities to other server worlds.
+ */
+public final class EntityTeleportationHelper {
```
Hmm, what's the point of Area Helper from minecraft then

i509VCB replied:
AreaHelper does not teleport the entity. The only teleport logic is Server player, entity moveToWorld and teleport command.

However an event for the after world change is gonna be a pain for other mod devs since they would have to go pretty manual. This just makes a lazy one method call to teleport an entity straight to another world.

LambdAurora replied:
I feel like this class should be in another module as it's more generalized than just for entity events.

i509VCB replied:
I believe that a method in updated dimension api pull request would suffice here and reduces duplicate logic.

liach reviewed (commented):
Api part look good, let me check impl later

valoeghese commented:
my fork of TooManyEvents (https://github.com/valoeghese/TooManyEvents) has a 1.16.2 EntitySpawnedCallback if you want to steal that

valoeghese commented:
:wink:

valoeghese commented:
Though it may be different to how fabric api wishes to implement. I did it in two parts: `PRE` and `POST`

valoeghese commented:
And it uses a Redirect

valoeghese commented:
So if you do implement it don't make it less powerful than the one in my fork of TooManyEvents because I don't want that to be incompatible with FAPI

i509VCB commented:
Valo, I may consider that for another day. The main thing is I don't feel I can do much more scope creep with this PR now.

shartte reviewed (commented):
LGTM, I am not sure about EntityTeleportationHelper being in here, I tried adding basically the same thing to the dimension-api PR

i509VCB committed:
Some renames, test events, drop damage event

Damage events need further consideration in future

i509VCB commented:
Damage event is no longer included due to some further considerations that are needed.

i509VCB committed:
Comments and null check

i509VCB committed:
Update for mappings

i509VCB committed:
Warning comment

LambdAurora reviewed (approved):
Always nice to have more events, but I currently don't know which events will trigger on which side, documentation stating this would be useful.

i509VCB committed:
Remove first join event and teleport helper

First join does not work atm and teleport is in dimensions pr

i509VCB commented:
Everything is pretty much ready for final reviews

i509VCB committed:
Module dependencies

i509VCB committed:
Javadoc tweaks

i509VCB committed:
Merge branch '1.16' into entity-post-damage-event

i509VCB committed:
The serverening

Player3324 reviewed (approved):
LGTM, thanks!

liach commented on fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/mixin/entity/event/LivingEntityMixin.java:
```diff
@@ -0,0 +1,38 @@
+
+@Mixin(LivingEntity.class)
+abstract class LivingEntityMixin extends EntityMixin {
+	@Inject(method = "onDeath", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/Entity;onKilledOther(Lnet/minecraft/server/world/ServerWorld;Lnet/minecraft/entity/LivingEntity;)V", shift = At.Shift.AFTER), locals = LocalCapture.CAPTURE_FAILEXCEPTION)
```
Imo we should just capture failhard than generate an exception stub. same for other local captures.
